src/gtest/test_checktransaction.cpp

```diff
@@ -391,3 +391,107 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
     EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-txns-invalid-joinsplit-signature", false)).Times(1);
     CheckTransactionWithoutProofVerification(tx, state);
 }
+
+TEST(checktransaction_tests, overwinter_constructors) {
+    CMutableTransaction mtx;
+    mtx.fOverwinter = true;
+    mtx.nVersion = 3;
+    mtx.nFormatBranchId =0xDEADBEEF;
+    mtx.nExpiryHeight = 20;
+
+    // Check constructor with overwinter fields
+    CTransaction tx(mtx);
+    EXPECT_EQ(tx.nVersion, mtx.nVersion);
+    EXPECT_EQ(tx.fOverwinter, mtx.fOverwinter);
+    EXPECT_EQ(tx.nFormatBranchId, mtx.nFormatBranchId);
+    EXPECT_EQ(tx.nExpiryHeight, mtx.nExpiryHeight);
+
+    // Check constructor of mutable transaction struct
+    CMutableTransaction mtx2(tx);
+    EXPECT_EQ(mtx2.nVersion, mtx.nVersion);
+    EXPECT_EQ(mtx2.fOverwinter, mtx.fOverwinter);
+    EXPECT_EQ(mtx2.nFormatBranchId, mtx.nFormatBranchId);
+    EXPECT_EQ(mtx2.nExpiryHeight, mtx.nExpiryHeight);
+    EXPECT_TRUE(mtx2.GetHash() == mtx.GetHash());
+
+    // Check assignment of overwinter fields
+    CTransaction tx2 = tx;
+    EXPECT_EQ(tx2.nVersion, mtx.nVersion);
+    EXPECT_EQ(tx2.fOverwinter, mtx.fOverwinter);
+    EXPECT_EQ(tx2.nFormatBranchId, mtx.nFormatBranchId);
+    EXPECT_EQ(tx2.nExpiryHeight, mtx.nExpiryHeight);
+    EXPECT_TRUE(tx2 == tx);
+}
+
+TEST(checktransaction_tests, overwinter_serialization) {
+    CMutableTransaction mtx;
+    mtx.fOverwinter = true;
+    mtx.nVersion = 3;
+    mtx.nFormatBranchId =0xDEADBEEF;
```

Comment:
Given that the serialization will check this value:

```
mtx.nFormatBranchId = NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId;
```

Reply:
I rebased.  I'm keeping 0xDEADBEEF for now since UPGRADE_OVERWINTER doesn't exist right now.  Marked as FIXME in the code.